services/dis-promrulegroups-operator/internal/controller/prometheusrule_controller.go

```diff
@@ -0,0 +1,451 @@
+	return r.handleCreation(ctx, req, prometheusRule)
+}
+
+func (r *PrometheusRuleReconciler) updateAnnotations(ctx context.Context, promRule monitoringv1.PrometheusRule, groupNames, armTemplateHash, armDeploymentName, timestamp string) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	annotations := promRule.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations[azPrometheusRuleGroupResourceNamesAnnotation] = groupNames
+	annotations[azArmTemplateHashAnnotation] = armTemplateHash
+	annotations[azArmDeploymentNameAnnotation] = armDeploymentName
+	annotations[azArmDeploymentLastSuccessfulTimestampAnnotation] = timestamp
+
+	promRule.SetAnnotations(annotations)
+	err := r.Client.Update(ctx, &promRule)
+	if err != nil {
+		logger.Error(err, "failed to update the PrometheusRule CR with new annotations", "namespace", promRule.Namespace, "name", promRule.Name)
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *PrometheusRuleReconciler) deployArmTemplate(ctx context.Context, deploymentName string, jsonTemplate string, actionGroupId string) error {
+	logger := log.FromContext(ctx)
+
+	contents := make(map[string]interface{})
+	_ = json.Unmarshal([]byte(jsonTemplate), &contents)
+
+	deploy, err := r.DeploymentClient.BeginCreateOrUpdate(
+		ctx,
+		r.AzResourceGroupName,
+		deploymentName,
+		armresources.Deployment{
+			Properties: &armresources.DeploymentProperties{
+				Template: contents,
+				Mode:     to.Ptr(armresources.DeploymentModeIncremental),
+				Parameters: map[string]interface{}{
+					"location": map[string]string{
+						"value": r.AzResourceGroupLocation,
+					},
+					"actionGroupId": map[string]string{
+						"value": actionGroupId,
+					},
+					"azureMonitorWorkspace": map[string]string{
+						"value": r.AzAzureMonitorWorkspace,
+					},
+					"clusterName": map[string]string{
+						"value": r.AzClusterName,
+					},
+				},
+			},
+		},
+		nil,
+	)
+
+	if err != nil {
+		logger.Error(err, "failed BeginCreateOrUpdate", "deploymentName", deploymentName)
+		return err
+	}
+	// TODO: Check the best practices here. I doubt we want to do this synchronously.
+	// From my tests it usually takes less than 5s tho so might actually be ok.
+	_, err = deploy.PollUntilDone(ctx, &azcoreruntime.PollUntilDoneOptions{Frequency: 5 * time.Second})
+	if err != nil {
+		return fmt.Errorf("cannot get the create deployment future respone: %v", err)
+	}
+	return nil
+}
+func (r *PrometheusRuleReconciler) deleteExternalResources(ctx context.Context, promRule monitoringv1.PrometheusRule) error {
+	allResourcesDeleted := true
+	annotations := promRule.GetAnnotations()
+	if resourceNames, ok := annotations[azPrometheusRuleGroupResourceNamesAnnotation]; ok {
+		for _, rn := range strings.Split(resourceNames, ",") {
+			if _, err := r.deletePrometheusRuleGroup(ctx, rn); err != nil {
+				allResourcesDeleted = false
```

Comment:
Should the error be logged to ease the debugging of why the delete failed?

Reply:
It's being logged in the function been called.